File: experiments/cjsawk/hex2.c
/* Copyright (C) 2016,2017 Jeremiah Orians
 * Copyright (C) 2017 Jan Nieuwenhuizen <janneke@gnu.org>
 * This file a mix of files from:
 *   mescc-tools
 *   M2-Planet
 *
 * with modifications (C) 2025 Liam Wilson
 *
 * mescc-tools is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * mescc-tools is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this file.  If not, see <http://www.gnu.org/licenses/>.
 */

#include <stdlib.h>
#include <sys/stat.h>

/* Globals */
int output;
struct entry** jump_tables;
int Base_Address;
int ip;
char* scratch;
char* filename;
int linenumber;

/* For processing bytes */
int hold;
int toggle;

char* source_filename;

#define max_string 4096
#define TRUE 1
#define FALSE 0

int EOF=-1;

struct input_files
{
	struct input_files* next;
	char* filename;
};

struct entry
{
	struct entry* next;
	unsigned target;
	char* name;
};

int match(char* a, char* b) {
	if((NULL == a) && (NULL == b)) return TRUE;
	if(NULL == a) return FALSE;
	if(NULL == b) return FALSE;

	int i = -1;
	do
	{
		i = i + 1;
		if(a[i] != b[i])
		{
			return FALSE;
		}
	} while((0 != a[i]) && (0 !=b[i]));
	return TRUE;
}


int in_set(int c, char* s) {
	/* NULL set is always false */
	if(NULL == s) return FALSE;

	while(0 != s[0]) {
		if(c == s[0]) return TRUE;
		s = s + 1;
	}
	return FALSE;
}

int consume_token(int source_file) {
	int i = 0;
	int c = fgetc(source_file);
	while(!in_set(c, " \t\n>")) {
		scratch[i] = c;
		i = i + 1;
		c = fgetc(source_file);
		if(EOF == c) {
			break;
		}
	}
	return c;
}

int Throwaway_token(int source_file) {
	int c;
	do
	{
		c = fgetc(source_file);
		if(EOF == c) {
			break;
		}
	} while(!in_set(c, " \t\n>"));

	return c;
}

int length(char* s) {
	int i = 0;
	while(0 != s[i]) {
		i = i + 1;
	}
	return i;
}

void Clear_Scratch(char* s) {
	do {
		s[0] = 0;
		s = s + 1;
	} while(0 != s[0]);
}

void Copy_String(char* a, char* b) {
	while(0 != a[0]) {
		b[0] = a[0];
		a = a + 1;
		b = b + 1;
	}
}

int GetHash(char* s) {
	int i = 5381;
	while(0 != s[0]) {
		i = i * 31 + s[0];
		s = s + 1;
	}
	return (i & 0xFFFF);
}

unsigned GetTarget(char* c) {
	struct entry* i;
	for(i = jump_tables[GetHash(c)]; NULL != i; i = i->next) {
		if(match(c, i->name)) {
			return i->target;
		}
	}
	exit(EXIT_FAILURE);
}

int storeLabel(int source_file, int ip) {
	struct entry* entry = calloc(1, sizeof(struct entry));

	/* Ensure we have target address */
	entry->target = ip;

	/* Store string */
	int c = consume_token(source_file);
	entry->name = calloc(length(scratch) + 1, sizeof(char));
	Copy_String(scratch, entry->name);
	Clear_Scratch(scratch);

	/* Prepend to list */
	int h = GetHash(entry->name);
	entry->next = jump_tables[h];
	jump_tables[h] = entry;

	return c;
}

void outputPointer(int displacement, int number_of_bytes) {
	unsigned byte;
	unsigned value = displacement;
	while(number_of_bytes > 0) {
		byte = value % 256;
		value = value / 256;
		fputc(byte, output);
		number_of_bytes = number_of_bytes - 1;
	}
}

void Update_Pointer(char ch) {
	/* Calculate pointer size*/
	if(in_set(ch, "%&")) {
		ip = ip + 4; /* Deal with % and & */
	} else if('!' == ch) {
		ip = ip + 1; /* Deal with ! */
	} else {
		exit(EXIT_FAILURE);
	}
}

void storePointer(char ch, int source_file) {
	/* Get string of pointer */
	Clear_Scratch(scratch);
	Update_Pointer(ch);
	int base_sep_p = consume_token(source_file);

	/* Lookup token */
	int target = GetTarget(scratch);
	int displacement;

	int base = ip;

	/* Change relative base address to :<base> */
	if ('>' == base_sep_p) {
		Clear_Scratch(scratch);
		consume_token (source_file);
		base = GetTarget (scratch);
	}
	displacement = (target - base);

	/* output calculated difference */
	if('!' == ch) {
		outputPointer(displacement, 1); /* Deal with ! */
	} else if('&' == ch) {
		outputPointer(target, 4); /* Deal with & */
	} else if('%' == ch) {
		outputPointer(displacement, 4);  /* Deal with % */
	} else {
		exit(EXIT_FAILURE);
	}
}

void line_Comment(int source_file) {
	int c = fgetc(source_file);
	while(!in_set(c, "\n\r")) {
		if(EOF == c) break;
		c = fgetc(source_file);
	}
	linenumber = linenumber + 1;
}

int hex(int c, int source_file) {
	if (in_set(c, "0123456789")) {
		return (c - 48);
	} else if (in_set(c, "abcdef")) {
		return (c - 87);
	} else if (in_set(c, "ABCDEF")) {
		return (c - 55);
	} else if (in_set(c, "#;")) {
		line_Comment(source_file);
	} else if ('\n' == c) {
		linenumber = linenumber + 1;
	}
	return -1;
}

void process_byte(char c, int source_file, int write) {
	if(0 <= hex(c, source_file)) {
		if(toggle) {
			if(write) {
				fputc(((hold * 16)) + hex(c, source_file), output);
			}
			ip = ip + 1;
			hold = 0;
		} else {
			hold = hex(c, source_file);
		}
		toggle = !toggle;
	}
}

void first_pass() {
	linenumber = 1;
	int source_file = fopen(source_filename, "r");

	toggle = FALSE;
	int c;
	for(c = fgetc(source_file); EOF != c; c = fgetc(source_file)) {
		/* Check for and deal with label */
		if(':' == c) {
			c = storeLabel(source_file, ip);
		}

		/* check for and deal with relative/absolute pointers to labels */
		if(in_set(c, "!%&")) {
			/* deal with 1byte pointer !; 4byte pointers (% and &) */
			Update_Pointer(c);
			c = Throwaway_token(source_file);
			if ('>' == c) { /* deal with label>base */
				c = Throwaway_token(source_file);
			}
		} else {
			process_byte(c, source_file, FALSE);
		}
	}
	fclose(source_file);
}

void second_pass() {
	linenumber = 1;
	int source_file = fopen(source_filename, "r");

	toggle = FALSE;
	hold = 0;

	int c;
	for(c = fgetc(source_file); EOF != c; c = fgetc(source_file)) {
		if(':' == c) {
			c = Throwaway_token(source_file);
		} else if(in_set(c, "!%&")) {
			storePointer(c, source_file);
		} else {
			process_byte(c, source_file, TRUE);
		}
	}
	fclose(source_file);
}

int main(int argc, char **argv) {
	jump_tables = calloc(65537, sizeof(struct entry*));
	Base_Address = 0x8048000;
	struct input_files* input = NULL;
	char* output_filename;
	scratch = calloc(max_string + 1, sizeof(char));
	struct input_files* temp;

	source_filename = argv[1];
	output_filename = argv[2];
	output = fopen(output_filename, "w");

	/* Get all of the labels */
	ip = Base_Address;
	first_pass();

	/* Fix all the references*/
	ip = Base_Address;
	second_pass();

	fclose(output);

	chmod(output_filename, 0750);

	return EXIT_SUCCESS;
}

File: experiments/cjsawk/hex2.js
function docs() {
  mks("This file must be valid C, JS, and AWK. See hex2_docs.txt");
  mks("AWK does not support C style comments so we can't use comments here.");
}

var output;
var jump_tables;
var Base_Address;
var ip;
var scratch;
var filename;
var linenumber;
var source_file;

var hold;
var toggle;

var source_filename;
var output_filename;

var max_string;

var TRUE;
var FALSE;

var EOF;
var NULL;

var entry_next_o;
var entry_target_o;
var entry_name_o;
var sizeof_entry;

function entry_next_g(o) {
	return ri32(add(o, entry_next_o));
}
function entry_next_s(o, v) {
	wi32(add(o, entry_next_o), v);
}

function entry_target_g(o) {
	return ri32(add(o, entry_target_o));
}
function entry_target_s(o, v) {
	wi32(add(o, entry_target_o), v);
}

function entry_name_g(o) {
	return ri32(add(o, entry_name_o));
}
function entry_name_s(o, v) {
	wi32(add(o, entry_name_o), v);
}

function ri32(o) {
	return OR(OR(AND(ri8(o), 255),
		shl(AND(ri8(add(o, 1)), 255), 8)),
		OR(shl(AND(ri8(add(o, 2)), 255), 16),
		shl(AND(ri8(add(o, 3)), 255), 24)));
}

function wi32(o, v) {
	wi8(o, AND(v, 255));
	v = shr(v, 8);
	wi8(add(o, 1), AND(v, 255));
	v = shr(v, 8);
	wi8(add(o, 2), AND(v, 255));
	v = shr(v, 8);
	wi8(add(o, 3), AND(v, 255));
}

function nextc() {
	return v_fgetc(source_file);
}

function smatch(a, b) {
	return smatch_(a, b, 0);
}
function smatch_(a, b,    i) {
	i = SUB(0, 1);
	while(1) {
		i = add(i, 1);
		if(neq(ri8(add(a, i)), ri8(add(b, i)))) {
			return FALSE;
		}
		if(AND(eq(0, ri8(add(a, i))), eq(0, ri8(add(b, i))))) {
			return TRUE;
		}
	}
}

function in_set(c, s) {
	while(neq(0, ri8(s))) {
		if(eq(c, ri8(s))) {
			return TRUE;
		}
		s = add(s, 1);
	}
	return FALSE;
}

function consume_token() {
	return consume_token_(0, 0);
}
function consume_token_(    i, c) {
	c = nextc();
	while(eq(0,in_set(c, mks(" \t\n>")))) {
		wi8(add(scratch, i), c);
		i = add(i, 1);
		c = nextc();
		if(eq(EOF, c)) {
			break;
		}
	}
	return c;
}

function Throwaway_token() {
	return Throwaway_token_(0);
}
function Throwaway_token_(c) {
	c = nextc();
	while(eq(0,in_set(c, mks(" \t\n>")))) {
		c = nextc();
		if(eq(EOF, c)) {
			break;
		}
	}
	return c;
}

function LENGTH(s) {
	return LENGTH_(s, 0);
}
function LENGTH_(s,    i) {
	while(neq(0, ri8(add(s, i)))) {
		i = add(i, 1);
	}
	return i;
}

function Clear_Scratch(s) {
	return Clear_Scratch_(s,0, 0);
}
function Clear_Scratch_(s,    i, c) {
	i = 0;
	c = ri8(add(s, i));
	while(neq(0, c)) {
		wi8(add(s, i), 0);
		i = add(i, 1);
		c = ri8(add(s, i));
	}
}

function Copy_String(a, b) {
	while(neq(0, ri8(a))) {
		wi8(b, ri8(a));
		a = add(a, 1);
		b = add(b, 1);
	}
}

function GetHash(s) {
	return GetHash_(s, 0);
}
function GetHash_(s,    i) {
	i = 5381;
	while(neq(0, ri8(s))) {
		i = add(shl(i, 5), add(i, ri8(s)));
		s = add(s, 1);
	}
	return AND(i, 65535);
}

function GetTarget(c) {
	return GetTarget_(c, 0);
}
function GetTarget_(c,    i) {
	i = ri32(add(jump_tables, mul(GetHash(c), 4)));
	while(neq(NULL, i)) {
		if(smatch(c, entry_name_g(i))) {
			return entry_target_g(i);
		}
		i = entry_next_g(i);
	}
	v_exit(1);
}

function storeLabel(ip) {
	return storeLabel_(ip, 0, 0, 0);
}
function storeLabel_(ip,    c, h, entry) {
	entry = v_calloc(1, sizeof_entry);

	entry_target_s(entry, ip);

	c = consume_token();
	entry_name_s(entry, v_calloc(add(LENGTH(scratch), 1), 1));
	Copy_String(scratch, entry_name_g(entry));
	Clear_Scratch(scratch);

	h = GetHash(entry_name_g(entry));
	entry_next_s(entry, ri32(add(jump_tables, mul(h, 4))));
	wi32(add(jump_tables, mul(h, 4)), entry);

	return c;
}

function outputPointer(displacement, number_of_bytes) {
	return outputPointer_(displacement, number_of_bytes, 0, 0);
}
function outputPointer_(displacement, number_of_bytes,    b, value) {
	value = displacement;
	while(gt(number_of_bytes, 0)) {
		b = mod(value, 256);
		value = AND(shr(value, 8), 16777215);
		fputc(b, output);
		number_of_bytes = SUB(number_of_bytes, 1);
	}
}

function Update_Pointer(ch) {
	if(in_set(ch, mks("%&"))) {
		ip = add(ip, 4);
	} else if(eq(mkC("!"), ch)) {
		ip = add(ip, 1);
	} else {
		v_exit(1);
	}
}

function storePointer(ch) {
	return storePointer_(ch, 0, 0, 0, 0);
}
function storePointer_(ch,    base_sep_p, base, target, displacement) {
	Clear_Scratch(scratch);
	Update_Pointer(ch);
	base_sep_p = consume_token();

	target = GetTarget(scratch);
	displacement;

	base = ip;

	if (eq(mkC(">"), base_sep_p)) {
		Clear_Scratch(scratch);
		consume_token();
		base = GetTarget(scratch);
	}
	displacement = SUB(target, base);

	if(eq(mkC("!"), ch)) {
		outputPointer(displacement, 1);
	} else if(eq(mkC("&"), ch)) {
		outputPointer(target, 4);
	} else if(eq(mkC("%"), ch)) {
		outputPointer(displacement, 4);
	} else {
		v_exit(1);
	}
}

function line_Comment() {
	return line_Comment_(0);
}
function line_Comment_(    c) {
	c = nextc();
	while(eq(0, in_set(c, mks("\n\r")))) {
		if(eq(EOF, c)) {
			break;
		}
		c = nextc();
	}
	linenumber = add(linenumber, 1);
}

function hex(c) {
	if (in_set(c, mks("0123456789"))) {
		return SUB(c, 48);
	} else if (in_set(c, mks("abcdef"))) {
		return SUB(c, 87);
	} else if (in_set(c, mks("ABCDEF"))) {
		return SUB(c, 55);
	} else if (in_set(c, mks("#;"))) {
		line_Comment();
	} else if (eq(mkC("\n"), c)) {
		linenumber = add(linenumber, 1);
	}
	return SUB(0, 1);
}

function process_byte(c, write) {
	if(lte(0, hex(c))) {
		if(toggle) {
			if(write) {
				fputc(add(mul(hold, 16), hex(c)), output);
			}
			ip = add(ip, 1);
			hold = 0;
		} else {
			hold = hex(c);
		}
		if(toggle) {
			toggle = FALSE;
		} else {
			toggle = TRUE;
		}
	}
}

function first_pass() {
	return first_pass_(0);
}
function first_pass_(    c) {
	linenumber = 1;
	source_file = v_fopen(source_filename, mks("r"));

	toggle = FALSE;
	c = nextc();
	while(neq(EOF, c)) {
		if(eq(mkC(":"), c)) {
			c = storeLabel(ip);
		}
		if(in_set(c, mks("!%&"))) {
			Update_Pointer(c);
			c = Throwaway_token();
			if (eq(mkC(">"), c)) {
				c = Throwaway_token();
			}
		} else {
			process_byte(c, FALSE);
		}
		c = nextc();
	}
	v_fclose(source_file);
}

function second_pass() {
	return second_pass_(0);
}
function second_pass_(    c) {
	linenumber = 1;
	source_file = v_fopen(source_filename, mks("r"));

	toggle = FALSE;
	hold = 0;

	c = nextc();
	while(neq(EOF, c)) {
		if(eq(mkC(":"), c)) {
			c = Throwaway_token();
		} else if(in_set(c, mks("!%&"))) {
			storePointer(c);
		} else {
			process_byte(c, TRUE);
		}
		c = nextc();
	}
	v_fclose(source_file);
}

function init_globals() {
	max_string  = 4096;

	TRUE = 1;
	FALSE = 0;

	EOF = SUB(0, 1);
	NULL = 0;

	entry_next_o = 0;
	entry_target_o = 4;
	entry_name_o = 8;
	sizeof_entry = 12;
}

function main(argc, argv) {
	init_support();
	init_globals();

	jump_tables = v_calloc(65537, 4);
	Base_Address = 134512640;

	if(eq(argc, 4)) {
		mks("hack for xp_linux win32 support see xp_linux readme");
		Base_Address = SUB(12288, 5632);
	}
	if(eq(argc, 5)) {
		mks("2nd hack for xp_linux win32 support see xp_linux readme");
		Base_Address = add(4194304, SUB(12288, 5632));
	}

	scratch = v_calloc(add(max_string, 1), 1);

	source_filename = ri32(add(argv, 4));
	output_filename = ri32(add(argv, 8));
	output = v_fopen(output_filename, mks("w"));

	ip = Base_Address;
	first_pass();

	ip = Base_Address;
	second_pass();

	v_fclose(output);

	chmod(output_filename, 488);

	return 0;
}
